keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,252 @@
+request that is enabled for newer etcd versions, it should warn in the logs that
+there might be an etcd version mismatch and fallback to the etcd functionality
+supported by all etcd 3.x versions.
+
+Also, Due to etcd upgrades and downgrades, there is no way to automatically
+detect the etcd version is a way that is guaranteed to be always correct, so the
+administrator must be able to set the version to a desired version.
+
+In addition to etcd minor version detection, all features requiring features
+introduced at a specific etcd minor version will have feature gates and will go
+through the usual kubernetes stability levels promotions (alpha, beta, GA).
+
+### Leveraging the Progress Notify Mechanism
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve
+consistent reads from cache if the etcd version is 3.4 or higher
+higher.  For etcd 3.3-, we would continue to serve consistent reads from etcd.
+
+Guard this optmization behind `WatchCacheConsistentReads` feature gate, and only server consistent reads from cache if it is enabled.
+
+When an consistent read request is received and the watch cache is enabled:
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful progress notify solve this by sending an event that will the highest revision observed to the >= the current revision. The cache is then guaranteed to to sufficiently up-to-date to serve a request at the current revision.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough, we may not need to request progress notify events from etcd.
```

Comment:
FYI: https://github.com/kubernetes/kubernetes/pull/86769 is the base for experiments

Reply:
This can only be considered experimental, the etcd kubernetes cluster used may serve other clients which maintains a large number of watchers, this value cannot be modified in a production environment.

Reply:
Our current goal is for a KEP is to describe what we need to take this feature to alpha (defaulted to off/experimental), so I think that's fine for now.